pkg/resource/autonaming/strategy.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package autonaming
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/urn"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+)
+
+// autonamingStrategy is a strategy for autonaming a resource.
+type autonamingStrategy interface {
+	GetOptions(urn urn.URN, randomSeed []byte) (opts *plugin.AutonamingOptions, deleteBeforeCreate bool)
+}
+
+// defaultAutonaming is the default autonaming config, which is equivalent to
+// no custom autonaming.
+type defaultAutonaming struct{}
+
+// defaultAutonamingConfig is the default instance of defaultAutonaming.
+var defaultAutonamingConfig = defaultAutonaming{}
```

Comment:
Do we use this anywhere? Seems like we always create a new instance when we resolve config. TBH I'm wondering if you just drop the interface and have `type AutonamingStrategy func(...)`. Then most of the strategies become functions and you don't need the structs (a couple return functions but I don't _think_ that's terrible -- could be wrong though).

Reply:
In fact -- why do we have the strategy interface _and_ `type Autonamer interface` in deployment? I feel like we should delete this, just have `Autonamer` (which all the stuff in the file then implements), and we should define that here. Then, every type in this file implements that interface and you have a nice strategy pattern where the `globalAutonaming` (which maybe we should rename to `rootAutonaming` or `stackConfigAutonaming`, maybe?) is actually a compound implementation of the same interface.

Reply:
Yeah, that's a great insight! I had to refactor a bunch of things to avoid module import cycles but I like the way things turn out in the end - except for `StackContext` comment that I posted below. Thank you for the suggestion!